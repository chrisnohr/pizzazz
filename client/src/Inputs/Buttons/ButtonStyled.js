import styled from 'styled-components';
import * as style from '../../styleConstants';

export const InputStyled = styled.input`
  display: inline-block;
  vertical-align: middle;
  font-size: ${style.respText};
  font-weight: ${style.textWeight};
  padding: 3vw 6vw;
  margin-bottom: ${style.respSpacer};
  color: ${style.textColor};
  border: .5px solid ${style.borderColor};
  background-color: ${style.white};
  box-shadow: ${style.shadow};
  @media screen and (min-width: 445px) {
    margin-bottom: ${style.spacer};
    font-size: 1.1rem;
    padding: ${style.padding};
  };
  :focus {
    outline: 0;
    border: 1px solid ${style.blue};
  }
`;

export const ButtonStyled = styled(InputStyled)`
  border-radius: .3rem;
  text-align: center;
  :active {
    box-shadow: inset ${style.shadow};
  }
`;



