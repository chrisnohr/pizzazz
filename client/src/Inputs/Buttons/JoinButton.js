import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';

const JoinButton = ({ countdownValue, canJoin, joinGame }) => {
    if (countdownValue > 20 && canJoin) {
      return (
        <Button
          value="Join the current game"
          clickHandler={joinGame}
        />
      );
    }
    return null;
  };
  
  JoinButton.propTypes = {
    countdownValue: PropTypes.number.isRequired,
    canJoin: PropTypes.bool.isRequired,
    joinGame: PropTypes.func.isRequired,
  };
  
  export default JoinButton;