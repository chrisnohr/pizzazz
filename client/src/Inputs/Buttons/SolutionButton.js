import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';
import { getWordScore } from '../../gameFlow/solutions/findWinner';
import { letterValues } from '../../Constants';

const SolutionButton = ({ isValidIndex, scrabbleBoard, sendSolution }) => {
    if (isValidIndex) {
      const solution = scrabbleBoard.substring(0, isValidIndex);
      const potentialScore = getWordScore(solution, letterValues);
      const text = `Play ${solution.toUpperCase()} for ${potentialScore} points`;
      const clickHandler = () => sendSolution(solution, potentialScore, isValidIndex);
      return (
        <Button
          value={text}
          clickHandler={clickHandler}
        />
      );
    }
    return null;
  };
  
  SolutionButton.propTypes = {
    isValidIndex: PropTypes.number.isRequired,
    scrabbleBoard: PropTypes.string.isRequired,
    sendSolution: PropTypes.func.isRequired,
  };
  
  export default SolutionButton;