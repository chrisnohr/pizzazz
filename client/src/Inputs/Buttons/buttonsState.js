import { take, select } from 'redux-saga/effects';
import { getScreenName } from '../ScreenName/screenNameState';
import { SEND_NAME } from '../ScreenName/screenNameState';
import { START_GAME } from '../../gameFlow/gameFlowState';

export const SEND_SOLUTION = 'SEND_SOLUTION';
export const JOIN_GAME = 'JOIN_GAME';

export const canJoinReducer = (state = false, action) => {
  switch (action.type) {
    case SEND_NAME:
      return true;
    case START_GAME:
      return false;
    default:
      return state;
  }
};

export const joinGameAction = () => ({
  type: JOIN_GAME,
});
  

export const sendSolutionAction = solution => ({
  type: SEND_SOLUTION,
  solution,
});
  
export function* sendSolutionIO(socket) {
  while (true) {
    const action = yield take(SEND_SOLUTION);
    const { solution } = action;
    const name = yield select(getScreenName);
    socket.emit('sendSolution', { solution, name });
  }
}
  
  