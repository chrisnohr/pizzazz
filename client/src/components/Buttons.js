import React from 'react';
import PropTypes from 'prop-types';
import { getScore } from '../clientLogic/findWinner';
import { letterValues } from '../Constants';
import { ButtonStyled } from './styled/ScreenNameStyled';

export const Button = ({ clickHandler, value }) => (
  <ButtonStyled
    type="button"
    readOnly
    autoFocus
    onClick={clickHandler}
    value={value}
  />
);

Button.propTypes = {
  clickHandler: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

export const JoinButton = ({ countdownValue, canJoin, joinGame }) => {
  if (countdownValue > 20 && canJoin) {
    return (
      <Button
        value="Join the current game"
        clickHandler={joinGame}
      />
    );
  }
  return null;
};

JoinButton.propTypes = {
  countdownValue: PropTypes.number.isRequired,
  canJoin: PropTypes.bool.isRequired,
  joinGame: PropTypes.func.isRequired,
};

export const SolutionButton = ({ isValidIndex, scrabbleBoard, sendSolution }) => {
  if (isValidIndex) {
    const solution = scrabbleBoard.substring(0, isValidIndex);
    const potentialScore = getScore(solution, letterValues);
    const text = `Play ${solution.toUpperCase()} for ${potentialScore} points`;
    const clickHandler = () => sendSolution(solution, potentialScore, isValidIndex);
    return (
      <Button
        value={text}
        clickHandler={clickHandler}
      />
    );
  }
  return null;
};

SolutionButton.propTypes = {
  isValidIndex: PropTypes.number.isRequired,
  scrabbleBoard: PropTypes.string.isRequired,
  sendSolution: PropTypes.func.isRequired,
};
