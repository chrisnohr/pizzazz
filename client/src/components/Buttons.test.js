import React from 'react';
import { shallow } from 'enzyme';
import { Button, JoinButton, SolutionButton } from './Buttons';
import 'jest-styled-components';

describe('<Button />', () => {
  it('calls the clickHandler function on click', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((<Button clickHandler={mockCallBack} value="Go" />));
    wrapper.dive().find('StyledComponent').simulate('click');
    expect(mockCallBack).toHaveBeenCalledTimes(1);
  });
  it('matches the snapshot', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow(<Button clickHandler={mockCallBack} value="Go" />);
    expect(wrapper).toMatchSnapshot();
  })
});

describe('<JoinButton>', () => {
  it('renders a <Button /> if canJoin is true and countdownValue > 20', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin={true}
        joinGame={mockCallBack}
      />
    ));
    expect(wrapper.find(Button).exists()).toEqual(true);
  });
  it('renders a button with value Join the current game', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin={true}
        joinGame={mockCallBack}
        />
    ));
    expect(wrapper.dive().props().value).toEqual('Join the current game');
  })
  it('does not render a <Button /> if canJoin is false', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin={false}
        joinGame={mockCallBack}
      />
    ));
    expect(wrapper.find(Button).exists()).toEqual(false);
  })
  it('does not render a <Button /> if countdownValue <= 20', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((
      <JoinButton
        countdownValue={20}
        canJoin={true}
        joinGame={mockCallBack}
      />
    ));
    expect(wrapper.find(Button).exists()).toEqual(false);
  })

  it('matches the snapshot', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((
      <JoinButton
        countdownValue={21}
        canJoin={true}
        joinGame={mockCallBack}
      />
    ));
    expect(wrapper).toMatchSnapshot();
  })
})

describe('<SolutionButton />', () => {
  it('renders a <Button /> with the valid word and its score', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((
      <SolutionButton
        isValidIndex={7}
        scrabbleBoard="pizzazz"
        sendSolution={mockCallBack}
      />
    ));
    const wrapper2 = shallow((
      <SolutionButton
        isValidIndex={5}
        scrabbleBoard="pizza00"
        sendSolution={mockCallBack}
      />
    ));
    expect(wrapper.dive().props().value).toEqual('Play PIZZAZZ for 46 points');
    expect(wrapper2.dive().props().value).toEqual('Play PIZZA for 26 points');
  })
  it('does not render a <Button /> when validIndex is zero', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow((
      <SolutionButton
        isValidIndex={0}
        scrabbleBoard="pizarro"
        sendSolution={mockCallBack}
      />
    ));
    expect(wrapper.find(Button).exists()).toEqual(false);
  })
})
