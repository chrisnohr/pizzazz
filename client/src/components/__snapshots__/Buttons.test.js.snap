// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches the snapshot 1`] = `
<Styled(styled.input)
  autoFocus={true}
  onClick={[MockFunction]}
  readOnly={true}
  type="button"
  value="Go"
/>
`;

exports[`<JoinButton> matches the snapshot 1`] = `
<Button
  clickHandler={[MockFunction]}
  value="Join the current game"
/>
`;
